usr/lib/cca_stdll/cca_specific.c

```diff
@@ -1495,6 +1499,315 @@ static CK_RV cca_iterate_adapters(STDLL_TokData_t *tokdata,
+                   CCA_MKVP_LENGTH) == 0)
+            ssd->preferred_found = TRUE;
+        break;
+    default:
+        return CKR_FUNCTION_FAILED;;
+    }
+
+    rc = cca_get_adapter_serial_number(ssd->serialno);
+    if (rc != CKR_OK)
+        return CKR_OK; /* adapter may be offline */
+
+    ssd->card = card;
+    ssd->domain = domain;
+    ssd->found = TRUE;
+
+    return CKR_OK;
+}
+
+static enum cca_mk_type cca_mk_type_from_key_type(enum cca_token_type keytype)
+{
+    switch (keytype) {
+    case sec_des_data_key:
+        return CCA_MK_SYM;
+    case sec_aes_data_key:
+    case sec_aes_cipher_key:
+    case sec_hmac_key:
+        return CCA_MK_AES;
+    case sec_rsa_priv_key:
+    case sec_ecc_priv_key:
+        return CCA_MK_APKA;
+    default:
+        return -1;
+    }
+}
+
+static struct cca_mk_change_op *cca_mk_change_find_op_by_keytype(
+                                                   STDLL_TokData_t *tokdata,
+                                                   enum cca_token_type keytype)
+{
+    struct cca_private_data *cca_private = tokdata->private_data;
+    enum cca_mk_type mk_type = cca_mk_type_from_key_type(keytype);
+    unsigned int idx;
+
+    if (cca_mk_change_find_mkvp_in_ops(tokdata, mk_type, &idx) == NULL)
+        return NULL;
+
+    return &cca_private->mk_change_ops[idx];
+}
+
+static CK_RV cca_select_single_apqn(STDLL_TokData_t *tokdata,
+                                    struct cca_mk_change_op *mk_change_op,
+                                    CK_BBOOL prefer_new_mk,
+                                    enum cca_token_type keytype,
+                                    char *serialno, CK_BBOOL *prefered_selected)
+{
+    struct cca_private_data *cca_private = tokdata->private_data;
+    unsigned char rule_array[256] = { 0, };
+    long return_code, reason_code, rule_array_count, device_name_len;
+    unsigned char *device_name;
+    struct cca_select_single_data ssd = { 0 };
+    CK_RV rc;
+
+    ssd.mk_change_op = mk_change_op;
+    ssd.prefer_new_mk = prefer_new_mk;
+    ssd.mk_type = cca_mk_type_from_key_type(keytype);
+
+    rc = cca_iterate_adapters(tokdata, cca_select_single_apqn_cb, &ssd);
+    if (rc != CKR_OK)
+        return rc;
+
+    if (ssd.found == FALSE) {
+        TRACE_ERROR("No single CCA APQN found\n");
+        return CKR_DEVICE_ERROR;
+    }
+
+    TRACE_DEVEL("single APQN %02X.%04X (Serialno %s) selected\n",
+                ssd.card, ssd.domain, ssd.serialno);
+    TRACE_DEVEL("APQN with prefered MK found: %d\n", ssd.preferred_found);
+
+    if (prefered_selected != NULL)
+        *prefered_selected = ssd.preferred_found;
+
+    /*
+     * If neither DEV-ANY, nor DOM-ANY is specified, no need to allocate the
+     * adapter, it a single adapter/domain configuration anyway.
+     */
+    if (!cca_private->dev_any && !cca_private->dom_any)
+        goto done;
+
+    /* Allocate the adapter */
+    memcpy(rule_array, "SERIAL  ", 8);
```

Comment:
You could use the CCA_KEYWORD_SIZE constant here, and below

Reply:
fixed